### ISSUE TITLE
widgets.TextEntry broken by 44da1724539823acb8f40fe0b62b378b31e6a0e8

### ISSUE THREAD

**Issue by ilar** [pyglet/pyglet#823]
**Describe the bug**
In example/gui/widgets.py, TextEntry is broken- it does not show the caret or the text. From testing, this is solely a display issue- clicking the box and typing, then pressing enter to 'commit' will correctly update the label on the right. (Unrelated, but your mouse MUST be hovering over the textbox for your input to register, even if you have focus. Probably not the intended functionality.)

Messing with the fg_group and bg_group order shows that the issue is due to some kind of issue with Groups- if fg_group is set to 5 or higher, the text will appear above the white box. removing _outline also allows the text to be correctly shown. Interestingly enough, the actual values do not matter- bg_group can be 6 and fg_group can be 5 and fg_group will still render over bg_group. fg_group just needs to be 5 or higher to take priority.

In the discord, someone tracked the offending commit down to [44da172](https://github.com/pyglet/pyglet/commit/44da1724539823acb8f40fe0b62b378b31e6a0e8). TextEntry works before that commit and does not afterwards.
Looking at the commit, a lot of it simply re-ordered code, with the only change of real import being lines 423-457.

Of particular weirdness to me is that when you try to set a group, instead it makes a brand new group with the requested group as the new group's parent. Does group depth affect draw order? That may be causing the issue since it then ignores the explicitly set order in TextEntry, and probably elsewhere too.

While I've provided system information, discord has shown this is crossplatform (windows and linux, at minimum) and carries through various versions of python.

**System Information:**

```python -m pyglet.info                                            (05-03 03:09)
Platform
------------------------------------------------------------------------------
platform:  Linux-5.15.0-71-generic-x86_64-with-glibc2.35
release:   5.15.0-71-generic
machine:   x86_64

Python
------------------------------------------------------------------------------
implementation: CPython
sys.version: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]
sys.maxint: 9223372036854775807
os.getcwd(): /home/#####/projects/pyglet

pyglet
------------------------------------------------------------------------------
pyglet.version: 2.0.6
pyglet.compat_platform: linux
pyglet.__file__: /home/####/projects/pyglet/pyglet/__init__.py
pyglet.options['audio'] = ('xaudio2', 'directsound', 'openal', 'pulse', 'silent')
pyglet.options['debug_font'] = False
pyglet.options['debug_gl'] = True
pyglet.options['debug_gl_trace'] = False
pyglet.options['debug_gl_trace_args'] = False
pyglet.options['debug_gl_shaders'] = False
pyglet.options['debug_graphics_batch'] = False
pyglet.options['debug_lib'] = False
pyglet.options['debug_media'] = False
pyglet.options['debug_texture'] = False
pyglet.options['debug_trace'] = False
pyglet.options['debug_trace_args'] = False
pyglet.options['debug_trace_depth'] = 1
pyglet.options['debug_trace_flush'] = True
pyglet.options['debug_win32'] = False
pyglet.options['debug_input'] = False
pyglet.options['debug_x11'] = False
pyglet.options['shadow_window'] = True
pyglet.options['vsync'] = None
pyglet.options['xsync'] = True
pyglet.options['xlib_fullscreen_override_redirect'] = False
pyglet.options['search_local_libs'] = True
pyglet.options['win32_gdi_font'] = False
pyglet.options['headless'] = False
pyglet.options['headless_device'] = 0
pyglet.options['win32_disable_shaping'] = False
pyglet.options['dw_legacy_naming'] = False
pyglet.options['win32_disable_xinput'] = False
pyglet.options['com_mta'] = False
pyglet.options['osx_alt_loop'] = False

pyglet.window
------------------------------------------------------------------------------
display: <pyglet.canvas.xlib.XlibDisplay object at 0x7f4208905db0>
screens[0]: XlibScreen(display=<pyglet.canvas.xlib.XlibDisplay object at 0x7f4208905db0>, x=2560, y=0, width=2560, height=1440, xinerama=True)
screens[1]: XlibScreen(display=<pyglet.canvas.xlib.XlibDisplay object at 0x7f4208905db0>, x=0, y=0, width=2560, height=1440, xinerama=True)
screens[2]: XlibScreen(display=<pyglet.canvas.xlib.XlibDisplay object at 0x7f4208905db0>, x=5120, y=0, width=2560, height=1440, xinerama=True)
config['double_buffer'] = 1
config['stereo'] = 0
config['buffer_size'] = 24
config['aux_buffers'] = 0
config['sample_buffers'] = 0
config['samples'] = 0
config['red_size'] = 8
config['green_size'] = 8
config['blue_size'] = 8
config['alpha_size'] = 0
config['depth_size'] = 24
config['stencil_size'] = 0
config['accum_red_size'] = 0
config['accum_green_size'] = 0
config['accum_blue_size'] = 0
config['accum_alpha_size'] = 0
config['major_version'] = 3
config['minor_version'] = 3
config['forward_compatible'] = None
config['opengl_api'] = 'gl'
config['debug'] = None
context: XlibContext(id=139921589788704, share=XlibContext(id=139921588251488, share=None))

window.context._info
------------------------------------------------------------------------------
gl_info.get_version(): (4, 6)
gl_info.get_vendor(): AMD
gl_info.get_renderer(): AMD Radeon RX 6800 XT (navi21, LLVM 15.0.7, DRM 3.42, 5.15.0-71-generic)

pyglet.gl.glx_info
------------------------------------------------------------------------------
context.is_direct(): 1
glx_info.get_server_vendor(): SGI
glx_info.get_server_version(): 1.4
glx_info.get_server_extensions():
   GLX_ARB_context_flush_control
   GLX_ARB_create_context
   GLX_ARB_create_context_no_error
   GLX_ARB_create_context_profile
   GLX_ARB_create_context_robustness
   GLX_ARB_fbconfig_float
   GLX_ARB_framebuffer_sRGB
   GLX_ARB_multisample
   GLX_EXT_create_context_es_profile
   GLX_EXT_create_context_es2_profile
   GLX_EXT_fbconfig_packed_float
   GLX_EXT_framebuffer_sRGB
   GLX_EXT_get_drawable_type
   GLX_EXT_libglvnd
   GLX_EXT_no_config_context
   GLX_EXT_texture_from_pixmap
   GLX_EXT_visual_info
   GLX_EXT_visual_rating
   GLX_MESA_copy_sub_buffer
   GLX_OML_swap_method
   GLX_SGI_make_current_read
   GLX_SGI_swap_control
   GLX_SGIS_multisample
   GLX_SGIX_fbconfig
   GLX_SGIX_pbuffer
   GLX_SGIX_visual_select_group
   GLX_INTEL_swap_event
glx_info.get_client_vendor(): Mesa Project and SGI
glx_info.get_client_version(): 1.4
glx_info.get_client_extensions():
   GLX_ARB_context_flush_control
   GLX_ARB_create_context
   GLX_ARB_create_context_no_error
   GLX_ARB_create_context_profile
   GLX_ARB_create_context_robustness
   GLX_ARB_fbconfig_float
   GLX_ARB_framebuffer_sRGB
   GLX_ARB_get_proc_address
   GLX_ARB_multisample
   GLX_EXT_buffer_age
   GLX_EXT_create_context_es2_profile
   GLX_EXT_create_context_es_profile
   GLX_EXT_fbconfig_packed_float
   GLX_EXT_framebuffer_sRGB
   GLX_EXT_import_context
   GLX_EXT_no_config_context
   GLX_EXT_swap_control
   GLX_EXT_swap_control_tear
   GLX_EXT_texture_from_pixmap
   GLX_EXT_visual_info
   GLX_EXT_visual_rating
   GLX_ATI_pixel_format_float
   GLX_INTEL_swap_event
   GLX_MESA_copy_sub_buffer
   GLX_MESA_query_renderer
   GLX_MESA_swap_control
   GLX_NV_float_buffer
   GLX_OML_swap_method
   GLX_OML_sync_control
   GLX_SGIS_multisample
   GLX_SGIX_fbconfig
   GLX_SGIX_pbuffer
   GLX_SGIX_visual_select_group
   GLX_SGI_make_current_read
   GLX_SGI_swap_control
   GLX_SGI_video_sync
glx_info.get_extensions():
   GLX_ARB_context_flush_control
   GLX_ARB_create_context
   GLX_ARB_create_context_no_error
   GLX_ARB_create_context_profile
   GLX_ARB_create_context_robustness
   GLX_ARB_fbconfig_float
   GLX_ARB_framebuffer_sRGB
   GLX_ARB_get_proc_address
   GLX_ARB_multisample
   GLX_EXT_buffer_age
   GLX_EXT_create_context_es2_profile
   GLX_EXT_create_context_es_profile
   GLX_EXT_fbconfig_packed_float
   GLX_EXT_framebuffer_sRGB
   GLX_EXT_no_config_context
   GLX_EXT_swap_control
   GLX_EXT_swap_control_tear
   GLX_EXT_texture_from_pixmap
   GLX_EXT_visual_info
   GLX_EXT_visual_rating
   GLX_INTEL_swap_event
   GLX_MESA_copy_sub_buffer
   GLX_MESA_query_renderer
   GLX_MESA_swap_control
   GLX_OML_swap_method
   GLX_OML_sync_control
   GLX_SGIS_multisample
   GLX_SGIX_fbconfig
   GLX_SGIX_pbuffer
   GLX_SGIX_visual_select_group
   GLX_SGI_make_current_read
   GLX_SGI_swap_control
   GLX_SGI_video_sync

pyglet.media
------------------------------------------------------------------------------
audio driver: <pyglet.media.drivers.openal.adaptation.OpenALDriver object at 0x7f42089cc850>

pyglet.media.ffmpeg
------------------------------------------------------------------------------
FFmpeg version: 4.4.2-0ubuntu0.22.04.1+esm1

pyglet.media.drivers.openal
------------------------------------------------------------------------------
Library: <CDLL 'libopenal.so.1', handle 556628dbbb20 at 0x7f4208a7f520>
Version: 1.1
Extensions:
   ALC_ENUMERATE_ALL_EXT
   ALC_ENUMERATION_EXT
   ALC_EXT_CAPTURE
   ALC_EXT_DEDICATED
   ALC_EXT_disconnect
   ALC_EXT_EFX
   ALC_EXT_thread_local_context
   ALC_SOFT_device_clock
   ALC_SOFT_HRTF
   ALC_SOFT_loopback
   ALC_SOFT_output_limiter
   ALC_SOFT_pause_device

pyglet.input.wintab
------------------------------------------------------------------------------
WinTab not available.
```

**How To Reproduce**
Running examples/gui/widgets.py will show the issue.

**Comment by caffeinepills** [comment]
I have confirmed this issue through testing.

I have made some changes here: #825 for a possible fix.

Test out that branch and if people can confirm that fixes it, I can merge.

**Comment by ilar** [comment]
Fixes it on my machine.

**Comment by pushfoo** [comment]
It's fixed on mine as well :+1:

**Comment by caffeinepills** [comment]
Merged in [c0b7e77](https://github.com/pyglet/pyglet/commit/c0b7e7797b545f2790138c28773da0015241b7e0)

**Comment by shenjackyuanjie** [comment]
I think this need to be reopen
